docker rehaul: remove repo2docker and add gpu support

stephchen opened the pull request:
Fixes WB-7840

Description
-----------
- removes repo2docker
- adds CUDA support for gpus
- some smaller things:
    - builds docker image based on current python version if not launching from a previous wandb run
    - adds `--gpu` flag to CLI to enable building a gpu enabled image (TODO: if rerunning a wandb run previously on gpu, set gpu on as default)
    - TODO: I don't think buildx partial caching (eg if we change some but not all of requirements.txt) is working at the moment, looking into this

metrics:
- for a test image with (numpy, torch, wandb):
    - 203.7s build time from scratch
    - 2.42gb image size (2.55gb with gpu)
    - 8.2s build time when rebuilding identical image
- compare with master:
    - 380.3s build time from scratch (370s building base image in repo2docker, 10.3s building on top)
    - 4.22gb image size (plus another 3.91gb base image stored on system)
    - 3.1s build time when rebuilding identical image — this caching is also dependent entirely on the target project name which is brittle)

Testing
-------
- tested manually locally with a few different python versions, tested on gpu via gcp
- writing unit tests now

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

vanpelt reviewed (commented):
This is looking cool, a couple things to consider:

1. The base image is really important here.  When "gpu" is enabled we should default to the latest version of cuda and allow the user to specify what cuda version they want.  Of course a user can override this entirely with a custom baseimage or image.  I wonder if we can derive the cuda version from wandb-metadata.json or elsewhere...
2. We should consider standardizing around miniconda for our builds.  The benefits will be faster build times for users that have conda configs and the ability to use any python version we want without doing an entire rebuild.  We're currently using miniconda in our [wandb/local build](https://github.com/wandb/core/blob/master/onprem/local/Dockerfile#L95-L112).  

The cool thing about conda is that we can install any version of python we want.  Potentially we make early steps of the build always install the miniconda py39 distribution, but then create a conda environment later with the appropriate python version, i.e.

```
conda env create --name=wandb python=3.7 pip
pip install -r requirements.txt
```

The trick is ensuring the appropriate conda environment is activated by default, happy to discuss more here if that's interesting.

We definitely have a lot of users using conda today and I think that's growing.  It's especially popular for anyone in windows land and MLFlow has standardized on it.

stephchen commented:
@vanpelt yeah conda is a good idea, repo2docker basically built their pip stuff on top of conda infra I think — lemme see if I can sub that in easily

cuda versioning is a good note, I haven't really tested it but eg cuda/torch compatibility is definitely a common issue

stephchen commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,6 +46,205 @@ def validate_docker_installation() -> None:
+
+ENV PYTHONUNBUFFERED=1
+
+ENTRYPOINT {command_arr}
```
todo: args don't make it into here, fix this

KyleGoyette commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,6 +47,207 @@ def validate_docker_installation() -> None:
+# make sure `python` points at the right version
+RUN update-alternatives --install /usr/bin/python python /usr/bin/python{py_version} 1 \
+    && update-alternatives --install /usr/local/bin/python python /usr/bin/python{py_version} 1
+"""
```
We should probably keep these in a different file, docker_templates or something

stephchen replied:
sorta prefer keeping these here for now since they are only used in once place?

KyleGoyette replied:
Okay, thats fine

KyleGoyette commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,6 +47,207 @@ def validate_docker_installation() -> None:
+def get_current_python_version():
+    full_version = sys.version.split()[0].split(".")
+    major = full_version[0]
+    version = ".".join(full_version[:2]) if len(full_version) >= 2 else major + ".0"
```
Is this just dropping the + for the dev versions of python potentially?

stephchen replied:
dropping the + would be correct since this is getting the version mostly to get the right docker base image, and i don't think they have them for dev versions

KyleGoyette commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,6 +47,207 @@ def validate_docker_installation() -> None:
+            "Docker BuildX is not installed, for faster builds upgrade docker: https://github.com/docker/buildx#installing"
+        )
+        requirements_line = "RUN WANDB_DISABLE_CACHE=true "
+    requirements_line += "pip install -r requirements.txt"
```
Doesn't this drop the use of `_wandb_bootstrap`? What if the repo doesn't have a requirements.txt?

stephchen replied:
fixed and readded this in

KyleGoyette reviewed (commented):
Couple of questions where I'm not sure if we're doing the right thing especially around installing requirements. Also still needs some tests.

I don't think we should block this on CUDA version. We should allow users to specify, defaulting to the newest. We should be able to add the Cuda version to RunInfo and pull it down with the rest.

stephchen commented:
- added ability for user to specify cuda version
- added conda support with buildkit caching
- fixed existing bug that was breaking buildkit caching

stephchen commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,53 +40,273 @@ def validate_docker_installation() -> None:
+            )
+
+        if docker.is_buildx_installed():
+            prefix = "RUN --mount=type=cache,mode=0777,target=/root/.cache/pip"
```
note that because we are doing a multistage build and the requirement installation takes place in the first stage, we can drop the uid/gid requirements from the buildkit configuration and do all installs as root. all installed deps are copied over in an env to the actual docker image provided to the user, which contains the setup for the user/uid/gid

KyleGoyette commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,53 +41,288 @@ def validate_docker_installation() -> None:
+    """Fill in the Dockerfile templates for stage 2 of build. CPU version is built on python:slim, GPU
+    version is built on nvidia:cuda"""
+
+    python_base_image = "python:{}-slim-buster".format(py_version)  # slim for running
```
We could allow users to specify a base image using the docker key of the launch spec config. in case they want to directly specify another image as base

KyleGoyette replied:
We may also need a windows option for windows users.

KyleGoyette commented on wandb/sdk/launch/docker.py:
```diff
@@ -45,53 +41,288 @@ def validate_docker_installation() -> None:
+            wandb.termwarn(
+                "Docker BuildX is not installed, for faster builds upgrade docker: https://github.com/docker/buildx#installing"
+            )
+            prefix = "RUN WANDB_DISABLE_CACHE=true"
```
could factor out this warn and prefix to outside of the other if statements no?

KyleGoyette commented on wandb/sdk/launch/launch.py:
```diff
@@ -161,6 +163,10 @@ def run(
         config = {}
 
+    if "docker" in config:
+        docker_args.update(config["docker"])  # userprovided args override
```
Are docker_args unused now? It seems like this  and the following line could be removed?

stephchen replied:
in theory they are passed thru to the `docker run` command  via `get_full_command`

KyleGoyette replied:
Got it

KyleGoyette commented on wandb/sdk/launch/runner/local.py:
```diff
@@ -81,77 +81,34 @@ def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
             )
-            image_uri = launch_project.docker_image
-            _logger.info("Getting docker command...")
+
```
We should be able to apply the final build step ontop of a user provided image, where we would set the environment variables and inject the launch-metadata.json file. Then we can get rid of the above warning

KyleGoyette replied:
Agreed

KyleGoyette commented on wandb/sdk/launch/_project_spec.py:
```diff
@@ -370,6 +374,7 @@ def create_project_from_spec(launch_spec: Dict[str, Any], api: Api) -> LaunchPro
         launch_spec.get("overrides", {}),
         launch_spec.get("resource", "local"),
         launch_spec.get("resource_args", {}),
+        launch_spec.get("cuda", False),
```
When we pull the `wandb-metadata.json` file we can check that for a cuda version if this is false.

KyleGoyette reviewed (changes requested):
A few small changes and some more questions. This is getting pretty close.

vanpelt reviewed (approved):
So much HOTNESS 🔥 .  Nice work!  

I can imagine simplifying this build even further in the future.  [Micromamba](https://mamba.readthedocs.io/en/latest/user_guide/micromamba.html) looks really sick.  Instead of installing the specific python version directly in Ubuntu or Debian, we could let micromamba do it.  It can handle conda envs, virtual envs, pip installations and python installations all in a tiny package.  Their official [docker image](https://hub.docker.com/r/mambaorg/micromamba) is 35MB 😮

KyleGoyette commented on tests/fixtures/launch/launch_sagemaker_config.json:
```diff
@@ -16,7 +16,7 @@
             },
             "StoppingCondition": {
-                "MaxRuntimeInSeconds": "60"
+                "MaxRuntimeInSeconds": 60
```
ty

KyleGoyette commented on tests/tests_launch/test_launch_aws.py:
```diff
@@ -423,32 +426,6 @@ def test_aws_get_region_file_fail_no_file(runner, monkeypatch):
         )
 
 
-def test_aws_fail_build(
```
Why is this test being removed?

KyleGoyette commented on tox.ini:
```diff
@@ -40,7 +40,7 @@ commands_pre =
     mkdir -p test-results
     py{35,36,37,38,39,310}:     python -m pytest {env:CI_PYTEST_SPLIT_ARGS:} -n={env:CI_PYTEST_PARALLEL:4} --durations=20 --junitxml=test-results/junit.xml --cov-config=.coveragerc --cov --cov-report= --no-cov-on-fail {posargs:tests/}
-    pylaunch:               jupyter-repo2docker --no-run ./tests/fixtures/
+    ; pylaunch:               jupyter-repo2docker --no-run ./tests/fixtures/
```
The caching here is busted, this can be removed.

KyleGoyette commented on wandb/sdk/launch/utils.py:
```diff
@@ -227,12 +236,12 @@ def download_entry_point(
     entity: str, project: str, run_name: str, api: Api, dir: str
 ) -> Optional[str]:
-    metadata = api.download_url(
+    reqs = api.download_url(  # @@@
```
nit: Remove this comment?

KyleGoyette commented on wandb/sdk/launch/runner/aws.py:
```diff
@@ -341,6 +304,10 @@ def build_sagemaker_args(
 
+    # remove args that were passed in for launch but not passed to sagemaker
+    sagemaker_args.pop("region", None)
+    sagemaker_args.pop("profile", None)
```
nit, there's another arg popped on line 291 if we want to group them

KyleGoyette reviewed (approved):
nice!
